### ISSUE TITLE
Regression with pyright version 1.1.344

### ISSUE THREAD

**Issue by hamdanal** [microsoft/pyright#6864]
**Describe the bug**
This code snippet passes with older pyright versions including 1.1.343 but fails with the latest version.

**Code or Screenshots**

```python
from typing import Any, Hashable, Iterator, Protocol, Sequence, SupportsIndex, TypeVar, overload

_T_co = TypeVar("_T_co", covariant=True)

class SequenceNotStr(Protocol[_T_co]):
    @overload
    def __getitem__(self, index: SupportsIndex, /) -> _T_co: ...
    @overload
    def __getitem__(self, index: slice, /) -> Sequence[_T_co]: ...
    def __contains__(self, value: object, /) -> bool: ...
    def __len__(self) -> int: ...
    def __iter__(self) -> Iterator[_T_co]: ...
    def index(self, value: Any, /, start: int = 0, stop: int = ...) -> int: ...
    def count(self, value: Any, /) -> int: ...
    def __reversed__(self) -> Iterator[_T_co]: ...

def f(x: SequenceNotStr[Hashable]):
    ...

s: list[str] = [""]
f(s)  # error: Argument of type "list[str]" cannot be assigned to parameter "x" of type "SequenceNotStr[Hashable]" in function "f"
```

**Extra context**
This protocol is extracted from the pandas-stubs project where the error started showing when I updated pyright.
Mypy does not report an error here

**VS Code extension or command-line**
command line: pyright 1.1.344

**Comment by erictraut** [comment]
I don't think this is a bug in pyright. The change you're seeing is due to [a recent bug fix in pyright](https://github.com/microsoft/pyright/issues/6833). The type checking logic wasn't properly honoring positional-only parameters in certain circumstances.

The cause of the problem should be clear from pyright's full error message, but in case it's not, here's an analysis...

Typeshed defines the `list.index` method as follows:

```python
    def index(self, __value: _T, __start: SupportsIndex = 0, __stop: SupportsIndex = sys.maxsize) -> int: ...
```

This means all three parameters for `list.index` are position-only. The `list.index` method does not accept `value`, `start` or `stop` by name.

Your protocol definition for `SequenceNotStr` requires that a compatible class accept these three parameters by name. That means `list` is not compatible with the `SequenceNotStr` protocol.

To fix this, you need to change the `index` method in your protocol as follows:

```python
    def index(self, value: Any, start: int = 0, stop: int = ..., /) -> int:
        ...
```

**Comment by hamdanal** [comment]
Thank you, you are right. Sorry for the noise.
